packages/core/src/actions/providers/getProvider.ts

```diff
@@ -8,11 +8,18 @@ export type GetProviderArgs = {
 
 export type GetProviderResult<TProvider extends Provider = Provider> = TProvider
 
+const providers = new Map()
+
 export function getProvider<TProvider extends Provider = Provider>({
   chainId,
 }: GetProviderArgs = {}): GetProviderResult<TProvider> {
   const client = getClient<TProvider>()
-  if (chainId && typeof client.config.provider === 'function')
-    return client.config.provider({ chainId })
+  if (chainId && typeof client.config.provider === 'function') {
+    if (providers.get(chainId)) return providers.get(chainId)
+
+    const provider = client.config.provider({ chainId })
+    providers.set(chainId, provider)
```

Comment:
Caching on `chainId` seems fine for most use cases, but someone could technically change `client.config.provider` through `client.setState` and this map would be stale. Could [pull it into `client`](https://github.com/wagmi-dev/wagmi/blob/main/packages/core/src/client.ts#L305) and bust whenever `client.config.provider` changes.

Reply:
Good idea!